Conda package attempt

ilaflott opened the pull request:
this PR:
- adds `meta.yaml` and `build.sh` to define the `conda build` package recipe
- adds `environment.yml` defining fully functional `conda` environment inwhich `mkmf` should work
- puts all `bin/` executables under a nested `mkmf` "package" directory for `conda build` recipe compatibility
- adds a github workflow file to run `conda build` under github's CI/CD pipelines, `conda build .` `--no-anaconda-upload` version in `.github/workflows/build_conda.yml`, with uploading in `.github/workflows/publish_conda.yml`. run conditions identical to that of `fre-cli`
- removes some old unused files (`.travis.yaml`, `.gitlab-ci.yml`)
- adjust success condition of some tests, e.g. `list_paths prints version` regex reflect recent version strings
- adjusts how `t/t001-list_paths.sh` does set-up, now creates `t/src/file6.f90` symbolic link on-the-fly. removed at end of each of those tests
- adjusts how every test does setup- checks `PATH` and adjusts only if needed. above `file6.f90` link creation done on-the-fly for `t/t003-mkmf.sh` as well. 
- tests and package should work in the usual way with the usual level of functioning for all desired contexts

ilaflott committed:
ham-fisted packaging approach- put everything in nested mkmf folder, define build recipe in meta.yaml and relevant fields. define conda environment.

ilaflott committed:
build_conda workflow

ilaflott commented:
green check, yes. note failures in pipeline...

ilaflott committed:
move test dir up one level and adjust testing commands, source files for test step. add git at dependency to see if it keeps the pipeline happy.

ilaflott committed:
Merge branch 'main' into conda_package

ilaflott commented:
current testing state is odd.

In the pipeline, `t/t001-list_paths.sh` tests fail, tests #'s 1, 3, 5, 6, and 7 (see copied pipeline output). But, locally on my workstation, only the first test, `1 list_paths prints version` fails.
```
+ bats t/t001-list_paths.sh
+ echo 'failure guard :-('

not ok 1 list_paths prints version
# (in test file t/t001-list_paths.sh, line 19)
#   `[[ "$output" =~ ^list_paths\ [0-9]+\.[0-9]+$ ]]' failed

not ok 3 list_paths using default out file
# (in test file t/t001-list_paths.sh, line 32)
#   `[ "$(wc -l < path_names)" -eq 6 ]' failed


not ok 5 list_paths find files in t and test_* directories
# (in test file t/t001-list_paths.sh, line 44)
#   `[ "$(wc -l < path_names)" -eq 8 ]' failed

not ok 6 list_paths find specific files in t or test_* directories
# (in test file t/t001-list_paths.sh, line 51)
#   `[ "$(wc -l < path_names)" -eq 7 ]' failed

not ok 7 list_paths with specified out file
# (in test file t/t001-list_paths.sh, line 59)
#   `[ "$(wc -l < $outFileName)" = "6" ]' failed

```

Additionally, in the pipeline, all `t/t002-git-version-string.sh` tests fail, and yet locally on my workstation, all of these tests pass. I suppose i'm not entirely surprised by this because i've seen 
 github's `CI/CD` has been weird about calling `git` commands directly in the past. 
```
+ bats t/t002-git-version-string.sh
+ echo 'failure guard :-('
```

ilaflott committed:
remove git dependency, add verbose run flag to bats testing calls. remove most unneeded failure-guarding shell commands.

ilaflott committed:
update success condition in regex-based check for version-printing test t/t001-list_paths.sh

ilaflott committed:
remove PATH env var export from test- shouldnt be needed anymore. change approach to resolving binDir in t/t001-list_paths- should possibly keep the pipeline happier. include mkmf in source files for test step of conda build. more echos, switch spot of echo output to avoid adjusting status of last-called shell command.

ilaflott committed:
torturing the pipeline and the test to see if i can get more intel. the testing approach makes perfect sense locally- so why isnt the pipeline happy?

ilaflott committed:
comment out most commands to test build- focus on certain things...

ilaflott committed:
torture test more. somewhat suspicious of insonsistent comparisons between string and ints in test success condition checks

ilaflott committed:
another attempt... set shell variable to number of lines determined via cat and wc -l

ilaflott committed:
remove failure guard and add a "no-parallelize-within-files" flag... slowing this pipeline down for its own good!! (maybe...)

ilaflott committed:
that option doesnt work for bats. specify only ONE job. add -x for tracing.

ilaflott committed:
what are you doing pipeline????

ilaflott committed:
what are you doing pipeline????

ilaflott commented:
AHA! the failure in githhub's pipeline arises because the command there, for some reason, resolves the symbolic link file `file6.linked` in `t/src` to file `file6.f90` with one of the targeted extensions. 

a little more simply: locally, the test `list_paths using default out file` in `t/t001-list_paths.sh` should ignore a link and only find 6 files. In github CI/CD, the symbolic link is resolved and it finds 7 files resulting in failure.

ilaflott committed:
adjust and torture tests again

ilaflott committed:
adjust and torture tests again because i forgot to change an expected answer -_-

ilaflott committed:
lets see what we have won...

ilaflott committed:
ok- throwing a towel in to focus on other things

ilaflott commented:
relatively sure that once `conda build .` enters the `test` phase, the `source_files` key specifying `t/` likely copies the directory with a command ala `cp -L -r t/ <target_dir>`, and is de-referencing the link to `file6.f90`, changing the answers in the `list_files` tests. i adjusted the answers accordingly, but this is honestly bad practice. 

the `git` based tests in `t` still do not work in the pipeline but work fine for me locally.

ceblanton commented:
It looks great, thanks!

The `noaa-gfdl` channel is not needed for building, is it?

What do you think about throwing in the noaa-gfdl anaconda channel key and having this pipeline publish as well as build. Too soon?

It's odd that the tests needing adjusting (but then again I hadn't realized there were mkmf tests. :)

ilaflott commented:
> The `noaa-gfdl` channel is not needed for building, is it?
nope!



> What do you think about throwing in the noaa-gfdl anaconda channel key and having this pipeline publish as well as build. Too soon?
no we can do it- i'll maybe say it should stick with an alpha version tag for now... at most, beta


> It's odd that the tests needing adjusting (but then again I hadn't realized there were mkmf tests. :)
100% conda build is derefencing links found in specified source files for the testing step! idk what the best way to deal with that is and not lose true quality control.

ilaflott committed:
Update README.md

put conda-relevant things in README- adjust markdown, update dependencies discussed. put disclaimer+license at the bottom of the readme to emphasize usage, installation, featurres etc.

ilaflott committed:
Update README.md

update badge

ilaflott committed:
adjust tests back to answers obtained locally

ilaflott commented:
> > The `noaa-gfdl` channel is not needed for building, is it?
> 
> nope!

actually now that i think about i i'm not entirely sure what you mean by this- care to elaborate? 

 
> > It's odd that the tests needing adjusting (but then again I hadn't realized there were mkmf tests. :)
> 
> 100% conda build is derefencing links found in specified source files for the testing step! idk what the best way to deal with that is and not lose true quality control.

i thought of a way- `bats` refreshes the testing directory a lot- i can just recreate the link-file at testing time as part of the `setup`

ilaflott committed:
settle on local/conda build answers in t/t001-list_paths.sh

ilaflott committed:
remove link file that we will create on-the-fly

ilaflott committed:
make the link on-the-fly- the local answers should now work in the pipeline.

ilaflott committed:
t003 setup adjustment as well.

ilaflott commented:
This is ready.

The only thing i can think of is that to make the tests compatible w/ conda and the past-mode of usage is to adjust `PATH` accordingly after checking e.g. `which mkmf` success.

ilaflott committed:
add publish workflow

ilaflott committed:
adjust setup routines to work for local and conda cases

ilaflott committed:
adjust (which mkmf) check in bats setup. lets try the git test in the pipeline again.

ilaflott committed:
clean up commented out shell

ilaflott committed:
have the setup function in t/t002-git-version-string configure the empty repo for pipeline-compatibility

ilaflott commented:
All tests now pass without caveat in all contexts.

ceblanton commented on .github/workflows/build_conda.yml:
```diff
@@ -0,0 +1,27 @@
+        run: |
+          # append the reqd channels
+          conda config --append channels conda-forge
+          conda config --append channels noaa-gfdl
```
noaa-gfdl channel packages are not needed for building mkmf, so this line could be removed, right?

ilaflott replied:
OHHHHH now i know what you meant!

ceblanton commented on environment.yml:
```diff
@@ -0,0 +1,8 @@
+name: mkmf
+channels:
+  - conda-forge
+  - noaa-gfdl
```
This can be removed

ceblanton commented on meta.yaml:
```diff
@@ -0,0 +1,45 @@
+package:
+  name: mkmf
+  version: 'alpha'
```
What about using the tag version here, 2025.01?

ilaflott replied:
lets go for it

ilaflott replied:
i made it the `Jinja2` filter that grabs the `git` tag, and we need to make a new `git` tag too here yeah?

ilaflott replied:
forget the `Jinja` for now

ceblanton reviewed (changes requested):
Nice work, Ian. Bonus for the the documentation updates.

The templates still have to be installed somehow, right?

ilaflott committed:
remove noaa-gfdl channel where not needed. update tag

ilaflott commented:
> Nice work, Ian. Bonus for the the documentation updates.
> 
> The templates still have to be installed somehow, right?

RE: these templates- `mkmf` needs you to specify a template target as an argument- i think. are these templates just supposed to be "found"? are they expected to be "nearby" ?

ilaflott committed:
replace some_channel in README

ilaflott committed:
forget the Jinja for now

ceblanton commented:
It's odd that mkmf includes the code _and_ the templates for compiling FMS models on selected platforms, but they're been used together like this for 20 years.

I think it's a little more consistent to only conda install the mkmf and list-paths tools, but by tradition, we should probably include the templates in the conda package also.

CatalogBuilder does this.. I'm not sure how it works, but e.g. the templates are installed alongside the bin and lib files. @Ciheim could explain it. For proof though, the GFDL default template (gfdl_template.json) is installed in fre-cli:

```
c2b:/nbhome/fms/conda/envs/fre-test%>find . -name gfdl_template.json 

./lib/python3.9/site-packages/catalogbuilder/cats/gfdl_template.json
```

ilaflott commented:
> It's odd that mkmf includes the code _and_ the templates for compiling FMS models on selected platforms, but they're been used together like this for 20 years.

using shell to drive workflows for decades will incentivize this kind of approach. unfortunately, given the way conda packages function... do we really want to be doing things like this? targeting the `templates/` dir within a conda environment i guess like:
```
mkmf -t /home/Ian.Laflotte/conda/envs/mkmf_test/......mkmf/templates
```
maybe we should consider de-coupling the templates and the code? 


> I think it's a little more consistent to only conda install the mkmf and list-paths tools, but by tradition, we should probably include the templates in the conda package also.

i'm sympathetic, i wanna figure it out.




> CatalogBuilder does this.. I'm not sure how it works, but e.g. the templates are installed alongside the bin and lib files. @Ciheim could explain it. For proof though, the GFDL default template (gfdl_template.json) is installed in fre-cli:
> 
> ```
> c2b:/nbhome/fms/conda/envs/fre-test%>find . -name gfdl_template.json 
> 
> ./lib/python3.9/site-packages/catalogbuilder/cats/gfdl_template.json
> ```

It's not quite 1-1: `mkmf` is not a `python` package and there is no `pip` install procedure here.

ilaflott commented:
`catalogbuilder` is using it's own `pkg_dir` property: https://github.com/search?q=repo%3ANOAA-GFDL%2FCatalogBuilder%20gfdl_template&type=code

[i get the executables in the environment to $PREFIX/bin](https://github.com/NOAA-GFDL/mkmf/blob/616d2cf49232543c909cd238f92280be9957f330/meta.yaml#L10-L13), because `conda` adds that to one's `PATH` for you. I could add a similar step where i `cp mkmf/templates/* $PREFIX/lib` or something. but `fre-cli` might have to do that `pkg_dir` path wrangling on it's own, and adjust the `-t` option, making the directory resolving `fre make`'s problem to solve

ilaflott commented:
i think the ideal solution would be that 
`mkmf -t /absolute/or/relative/path/to/template.mk` works as it does now, but e.g. specifying `mkmf -t ncrc-gcc` without the `.mk` extension says to `mkmf` "hey, get this template from your `include` dir".

ceblanton commented:
Yes!! I agree that would be a natural solution. I was thinking of a dummy `list-templates` tool to be added that reports the template directory but your idea is better.

singhd789 commented:
> > Nice work, Ian. Bonus for the the documentation updates.
> > The templates still have to be installed somehow, right?
> 
> RE: these templates- `mkmf` needs you to specify a template target as an argument- i think. are these templates just supposed to be "found"? are they expected to be "nearby" ?
> 
> EDIT: https://github.com/NOAA-GFDL/fre-cli/blob/a9730e039274739e5dc8710d82dcc09578735d18/fre/make/gfdlfremake/buildDocker.py#L31

Pertaining to the templates, currently we have the path specified in the platforms.yaml:
https://github.com/NOAA-GFDL/fre-cli/blob/a9730e039274739e5dc8710d82dcc09578735d18/fre/make/tests/null_example/platforms.yaml#L6

Ryan mentioned this in another issue as well (in relation to mkmf being a submodule but still applies): https://github.com/orgs/NOAA-GFDL/projects/32?pane=issue&itemId=92869904&issue=NOAA-GFDL%7Cfre-cli%7C309

ilaflott commented:
My main problem with me implementing that in this PR is that i do not know `perl` and am unsure if i'm the best person to implement that functionality on a short time-scale.

ilaflott commented:
Let's chat for a minute in-person on this when we can

ilaflott commented:
The templates should be a separate repository, as they're not really used *by* this repository out of necessity- they are a convenience for prospective users.

We're informing this view point with the following: 

1) all `mkmf` does with the value given to `-t` is stick the value as-is into the output Makefile, and the template is not read nor it's contents used programmatically by `mkmf`. 
```
> mkmf -t FOO
. Makefile is ready.

> cat Makefile | grep TEMPLATE
MK_TEMPLATE = FOO
include $(MK_TEMPLATE)
```
2) no one to our knowledge is relying on the templates in this repo as being in a particular spot directory-structure wise.

ilaflott committed:
move the templates back to base of the repo directory, outside of the package directory.

ilaflott committed:
Update README.md

slightly adjust current README instructions

ilaflott committed:
Update publish_conda.yml

test-run publish_conda workflow by asking it to run on push to `conda_package` branch, see what happens!

ilaflott committed:
restore publish_conda workflow run condition. remove uneeded and old gitlab/travis pipeline workflow files

ilaflott committed:
Merge branch 'main' into conda_package

singhd789 commented on .github/workflows/build_conda.yml:
Just curious, `mkmf` isn't really touched that much I don't think...right? If it was at all, Mikyung had that really nice addition in the fre-cli build_conda script:
https://github.com/NOAA-GFDL/fre-cli/blob/main/.github/workflows/build_conda.yml#L8-L10

singhd789 commented on .github/workflows/publish_conda.yml:
```diff
@@ -0,0 +1,31 @@
+          conda install conda-build conda-verify
+
+          # set token and upload preference
+          export ANACONDA_API_TOKEN=${{ secrets.ANACONDA_TOKEN }}
```
For my own clarification, what were the token and secrets for again?

singhd789 commented on README.md:
```diff
@@ -1,59 +1,103 @@
+Usage as a `conda` package requires `conda`.
+
+For testing, we recommend [`bats-core`](https://github.com/bats-core/bats-core),
+a current and up-to-date (as of January 2025) fork of the original 
```
looks like there was a new release as of May 18th

singhd789 commented on README.md:
```diff
@@ -1,59 +1,103 @@
-Kushner's suggestion).
+Usage as a `conda` package requires `conda`.
+
+For testing, we recommend [`bats-core`](https://github.com/bats-core/bats-core),
```
Might be nice to include what bats stands for (Bash Automated Testing System)..because I had no idea what it was 😅

singhd789 replied:
and maybe a quick little blurb about why it's recommended?

singhd789 commented on README.md:
```diff
@@ -1,59 +1,103 @@
+git clone https://github.com/noaa-gfdl/mkmf.git mkmf && cd mkmf
+conda env create -y -f ./environment.yaml
+conda activate mkmf
+export $PATH=$PWD/mkmf/bin:$PATH
```
Wait, why would it be necessary to add it to the path if there is a conda package for mkmf?

ilaflott replied:
this is like the "developer install" of `fre-cli`- local code responds to local edits, so add the local `mkmf/bin` to `PATH`.

singhd789 commented on README.md:
A bit unrelated, but do you know if anyone has made a simplified conda packaging guide at the lab or is it really, just follow along with tutorials online?

singhd789 replied:
this is really good development, so I'm just curious

ceblanton replied:
Conda packaging is great candidate for the 5-10 minute MSD training topics, I agree.

I'm afraid yours truly attempted to do exactly this, building off of a Seth guide, as part of a conda MSD seminar a year or two ago, but the conda packaging parts were not that well-developed.

ilaflott replied:
i did send these notes around!

underwoo reviewed (changes requested):
Some changes, or additional information needed.

ilaflott committed:
Remove tcsh from requirements in meta.yaml

Removed 'conda-forge::tcsh' from host, run, and test requirements.

ilaflott committed:
Merge branch 'main' into conda_package

ilaflott commented on t/t000-basic.sh:
```diff
@@ -1,9 +1,17 @@
+   binDir=$(readlink -f ${BATS_TEST_DIRNAME}/../mkmf/bin)
+   do_we_have_mkmf=$(which mkmf) || echo "no we do not!"
+   if [ $do_we_have_mkmf ]; then
+	   echo 'likely conda case'
```
to your point, coming back after all this time and trying to recall why I put this here, a clarifying comment is indeed the minimum re-thought required here.

ilaflott replied:
i *think* this is resolveable, but sometimes the best practices with `PATH` elude me. feel free to unresolve if i'm wrong

ilaflott committed:
Install templates to $PREFIX/share/mkmf/templates; fix t004 template list (targeting #75 review)

meta.yaml:
- Add build steps to copy templates/* to $PREFIX/share/mkmf/templates/
  so the templates are part of the conda package and discoverable by
  downstream tools (e.g. fre-cli) at a stable, predictable path:
    mkmf -t $CONDA_PREFIX/share/mkmf/templates/<name>.mk

t/t004-templates.sh:
- setup(): detect conda context via $PREFIX env var and set templateDir
  to the installed location; fall back to in-source templates/ otherwise
- Remove stale template tests that no longer exist in templates/:
    ncrc-cray.mk, ncrc-gnu.mk, ncrc-pgi.mk, ncrc-intel.mk
- Add tests for all templates that currently exist in templates/:
    hpcme-intel2025.03-oneapi.mk, hpcme-intel21/23/24.mk,
    linux-intel.mk, ncrc-nvhpc.mk, ncrc4-cce/gcc/intel.mk,
    ncrc5-cce/gcc/intel/intel-classic/intel-oneapi.mk

ilaflott committed:
Address underwoo's review feedback on test setup/teardown hygiene

t000, t001, t002, t003, t004 — PATH check (underwoo line 8 / 'see above'):
- Replace fragile 'do_we_have_mkmf=$(which mkmf) || echo ...' pattern
  with 'if [ $(command -v mkmf) ]' as suggested
- Add comment explaining the dual-mode rationale: conda-installed copy is
  already on PATH and is the copy under test; local dev adds bin/ so the
  working copy is exercised instead of any system copy

t001 — symlink isolation (underwoo lines 14 and 24):
- setup(): create testDir first, copy t/src/ into it, create the
  file6.f90 symlink inside testDir/src/ instead of in the real source tree
  so teardown cleans it automatically and a killed test leaves no residue
- teardown(): remove the now-unnecessary 'rm -f .../src/file6.f90' line
- Update all test invocations from BATS_TEST_DIRNAME/src -> testDir/src

t001 — count_lines usage (underwoo line 44):
- Apply count_lines() helper to all seven 'cat file | wc -l' sites

t003 — symlink isolation (same class as t001):
- Same setup/teardown changes: copy src to testDir, symlink inside testDir
- Update all BATS_TEST_DIRNAME/src references in test bodies to testDir/src
  including path-matching regexes in 'mkmf finds sources in directory'
  and 'mkmf finds include files in -I<INCLUDE>'

t004 — fix syntax error from previous commit:
- Restore missing testDir creation, cd, and teardown() that were
  accidentally dropped; setup() was left unclosed causing bats parse error

ilaflott committed:
Remove tcsh from environment.yml; move bats-core to test-only in meta.yaml

- Remove tcsh dependency from environment.yml per underwoo's review
  (tcsh is not used in any mkmf tools)
- Remove bats-core from host: and run: requirements in meta.yaml
  (bats-core is a test runner, not a runtime dependency; it remains
  in test.requires where it belongs)

ilaflott committed:
Always prepend correct bin dir to PATH; never skip (underwoo review)

Address underwoo's follow-up review comment: the old approach skipped
the PATH prepend when mkmf was already found on PATH, which could
silently test a stale or unrelated copy and hide real failures.

New approach: always prepend the appropriate directory to PATH.
- In conda build test environments ($PREFIX set): use $PREFIX/bin
  where the just-built package was installed.
- In local development: use the repo's own mkmf/bin/ directory.

This ensures the tests always exercise the intended copy of mkmf
regardless of what else is on the user's PATH.

ilaflott committed:
test: fail loudly when mkmf is not on PATH instead of silent fallback

Replace the if/else pattern that silently fell back to the repo's
mkmf/bin/ directory with a two-step approach:

1. In conda build/test ($PREFIX set), prepend $PREFIX/bin to PATH
2. Verify mkmf is on PATH; if not, error out with an actionable message
   telling the user to add mkmf/bin to their PATH

This addresses underwoo's review feedback: the old else branch could
resolve a path that doesn't work and mask the real failure.  Now,
missing mkmf always causes an immediate, clear failure.

ilaflott committed:
ci: add non-conda test workflow

Runs bats tests directly against the repo checkout using system
perl and bats from apt, exercising the local-development PATH
setup in the test harness.

ilaflott committed:
ci: merge build_conda and publish_conda into build_and_publish_conda

On pull_request: build only (test).
On push to main: build and publish to anaconda.org using ANACONDA_TOKEN.
Removes the now-redundant publish_conda workflow.

ilaflott commented:
I decided that those testing locally checked out code will be yelled at by the tests if they have no added `mkmf`'s executables to `PATH` by hand. This case is covered by the new, non-`conda` oriented `test` workflow. `conda build` still covers the package use-case

edit: `.github/workflow/test.yml` not failing nor succeeding- stalling. why?

ilaflott committed:
test: fix t002 hanging on GitHub Actions

Replace 'tr < /dev/urandom | fold | head' pipelines with bounded
'dd bs=45 count=1 | base64' approach.  The old pipeline can hang
when SIGPIPE delivery to tr is delayed, which was observed on bare
ubuntu-latest runners (the conda container was unaffected).

ilaflott commented:
Looks like this is ready for round 2.

ilaflott commented:
https://github.com/NOAA-GFDL/fre-cli/pull/774 checks the compatibility of this with `fre-cli`'s current usage. should be usable
